src/parquet/column/levels-test.cc

```diff
@@ -0,0 +1,133 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#include <gtest/gtest.h>
+
+#include "parquet/thrift/parquet_types.h"
+#include "parquet/column/levels.h"
+
+using std::string;
+
+namespace parquet_cpp {
+
+class TestLevels : public ::testing::Test {
+ public:
+  void InitBuffers() {
+    input_levels_.resize(num_levels_);
+    output_levels_.resize(num_levels_);
+    bytes_.resize(2 * num_levels_);
+  }
+
+  int GenerateLevels(int min_repeat_factor, int max_repeat_factor) {
+    int total_count = 0;
+    // for each repetition count upto max_repeat_factor
+    for (int repeat = min_repeat_factor; repeat <= max_repeat_factor; repeat++) {
+      // repeat count increase by a factor of 2 for every iteration
+      int repeat_count = (1 << repeat);
+      // generate levels for repetition count upto the maximum level
+      int value = 0;
+      int bwidth = 0;
+      while (value <= max_level_) {
+        for (int i = 0; i < repeat_count; i++) {
+          input_levels_[total_count++] = value;
+        }
+        value = (2 << bwidth) - 1;
+        bwidth++;
+      }
+    }
+    return total_count;
+  }
+
+  void VerifyLevelsEncoding(parquet::Encoding::type encoding) {
+    LevelEncoder encoder;
+    LevelDecoder decoder;
+    int levels_count = 0;
+    // start encoding and decoding
+    if (encoding == parquet::Encoding::RLE) {
+      // leave space to write the rle length value
+      encoder.Init(encoding, max_level_, num_levels_,
+          bytes_.data() + sizeof(uint32_t), sizeof(uint8_t) * 2 * num_levels_);
+
+      levels_count = encoder.Encode(num_levels_, input_levels_.data());
+      (reinterpret_cast<uint32_t*>(bytes_.data()))[0] = encoder.len();
+
+    } else {
+      encoder.Init(encoding, max_level_, num_levels_,
+          bytes_.data(), sizeof(uint8_t) * 2 * num_levels_);
+      levels_count = encoder.Encode(num_levels_, input_levels_.data());
+    }
+
+    ASSERT_EQ(levels_count, num_levels_);
+
+    decoder.Init(encoding, max_level_, num_levels_, bytes_.data());
+    levels_count = decoder.Decode(num_levels_, output_levels_.data());
+
+    ASSERT_EQ(levels_count, num_levels_);
+
+    for (int i = 0; i < num_levels_; i++) {
+      EXPECT_EQ(input_levels_[i], output_levels_[i]);
+    }
+  }
+
+  void SetUp() {}
+
+  void TearDown() {}
+
+ protected:
+  std::vector<int16_t> input_levels_;
+  std::vector<int16_t> output_levels_;
+  std::vector<uint8_t> bytes_;
+  int max_level_;
+  int num_levels_;
+};
```

Comment:
I added `max_level_` and `num_levels_` as class variables since multiple methods require them.